AssertionError: Dynamo only supports FSDP with use_orig_params=True

### System Info

cuda 11.7
accelerate=0.21.0.dev0
transformers=4.31.0.dev0
torch=2.0.1
python=3.8


### Who can help?

@pacman100 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

### command:
accelerate launch --config_file accelerate_config.yaml --num_machines 7 --num_processes 28 --machine_rank $NODE_RANK --main_process_ip $MASTER_ADDR --main_process_port $MASTER_PORT ./trainer.py --model_name_or_path ".."  --data_path ".." --per_device_train_batch_size 24 --per_device_eval_batch_size 24 --do_train --evaluation_strategy no --output_dir outputs --learning_rate 2e-5 --num_train_epochs 4 --lr_scheduler_type cosine --warmup_ratio 0.03 --weight_decay 0.0 --logging_steps 1 --save_strategy epoch --bf16 true --tf32 true --load_best_model_at_end false --model_max_length 2048 --gradient_checkpointing true --save_total_limit 1 --model_resume_from_checkpoint false --torch_compile true


### accelerate_config.yaml

```
compute_environment: LOCAL_MACHINE
distributed_type: FSDP
downcast_bf16: 'no'
fsdp_config:
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_backward_prefetch_policy: BACKWARD_PRE
  fsdp_forward_prefetch: false
  fsdp_offload_params: false
  fsdp_sharding_strategy: 1
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_sync_module_states: true
  fsdp_transformer_layer_cls_to_wrap: LlamaDecoderLayer
  fsdp_use_orig_params: true
main_training_function: main
num_machines: 1
num_processes: 2
mixed_precision: bf16
rdzv_backend: static
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

### basic training code
```
def train():
    print("Env Variables")
    env_vars = os.environ
    for key, value in env_vars.items():
        print(key, "=", value)
    parser = transformers.HfArgumentParser(
        (ModelArguments, DataArguments, TrainingArguments))
    model_args, data_args, training_args = parser.parse_args_into_dataclasses()
    model = transformers.LlamaForCausalLM.from_pretrained(
        model_args.model_name_or_path,
        cache_dir=training_args.cache_dir,
    )

    tokenizer = transformers.AutoTokenizer.from_pretrained(
        model_args.model_name_or_path,
        cache_dir=training_args.cache_dir,
        model_max_length=training_args.model_max_length,
        padding_side="right"
    )
    
    if tokenizer.pad_token is None:
        smart_tokenizer_and_embedding_resize(
            tokenizer=tokenizer,
            model=model,
        )


    data_module = make_hf_data_module(tokenizer=tokenizer,
                                              data_args=data_args)
    trainer = Trainer(model=model,
                    tokenizer=tokenizer,
                    args=training_args,
                    **data_module)
    
    if model_args.model_resume_from_checkpoint:
        trainer.train(resume_from_checkpoint=model_args.model_name_or_path)
    else:
        trainer.train()
    trainer.save_state()
    safe_save_model_for_hf_trainer(trainer=trainer,
                                   output_dir=training_args.output_dir)
```

###Stacktrace

  

> You can suppress this exception and fall back to eager by setting:
>   torch._dynamo.config.suppress_errors = True
>   self.symbolic_locals = collections.OrderedDict(
>   File "/opt/conda/envs/ptca/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 1673, in <genexpr>
>     self.symbolic_locals = collections.OrderedDict(
>   File "/opt/conda/envs/ptca/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 1673, in <genexpr>
>     VariableBuilder(
>   File "/opt/conda/envs/ptca/lib/python3.8/site-packages/torch/_dynamo/variables/builder.py", line 172, in __call__
>     return self._wrap(value).clone(**self.options())
>   File "/opt/conda/envs/ptca/lib/python3.8/site-packages/torch/_dynamo/variables/builder.py", line 248, in _wrap
>     VariableBuilder(
>   File "/opt/conda/envs/ptca/lib/python3.8/site-packages/torch/_dynamo/variables/builder.py", line 172, in __call__
>     output = [
>   File "/opt/conda/envs/ptca/lib/python3.8/site-packages/torch/_dynamo/variables/builder.py", line 249, in <listcomp>
>     return self._wrap(value).clone(**self.options())
>   File "/opt/conda/envs/ptca/lib/python3.8/site-packages/torch/_dynamo/variables/builder.py", line 248, in _wrap
>     VariableBuilder(self.tx, GetItemSource(self.get_source(), i))(
>   File "/opt/conda/envs/ptca/lib/python3.8/site-packages/torch/_dynamo/variables/builder.py", line 172, in __call__
>     output = [
>   File "/opt/conda/envs/ptca/lib/python3.8/site-packages/torch/_dynamo/variables/builder.py", line 249, in <listcomp>
>     return self._wrap(value).clone(**self.options())
>   File "/opt/conda/envs/ptca/lib/python3.8/site-packages/torch/_dynamo/variables/builder.py", line 345, in _wrap
>     VariableBuilder(self.tx, GetItemSource(self.get_source(), i))(
>   File "/opt/conda/envs/ptca/lib/python3.8/site-packages/torch/_dynamo/variables/builder.py", line 172, in __call__
>     assert getattr(
> AssertionError: Dynamo only supports FSDP with use_orig_params=True
> 
> Set torch._dynamo.config.verbose=True for more information

### Expected behavior

torch.compile works smoothly with FSDP

--- a/src/transformers/trainer.py
+++ b/src/transformers/trainer.py
@@ -3836,8 +3836,10 @@ class Trainer:
         # post accelerator creation setup
         if self.is_fsdp_enabled:
             fsdp_plugin = self.accelerator.state.fsdp_plugin
-            fsdp_plugin.limit_all_gathers = self.args.fsdp_config.get("limit_all_gathers", False)
-            fsdp_plugin.use_orig_params = self.args.fsdp_config.get("use_orig_params", False)
+            fsdp_plugin.limit_all_gathers = self.args.fsdp_config.get(
+                "limit_all_gathers", fsdp_plugin.limit_all_gathers
+            )
+            fsdp_plugin.use_orig_params = self.args.fsdp_config.get("use_orig_params", fsdp_plugin.use_orig_params)
 
         if self.is_deepspeed_enabled:
             if getattr(self.args, "hf_deepspeed_config", None) is None:

--- a/src/transformers/trainer.py
+++ b/src/transformers/trainer.py
@@ -56,7 +56,7 @@ from . import __version__
 from .configuration_utils import PretrainedConfig
 from .data.data_collator import DataCollator, DataCollatorWithPadding, default_data_collator
 from .debug_utils import DebugOption, DebugUnderflowOverflow
-from .deepspeed import deepspeed_init, deepspeed_load_checkpoint, is_deepspeed_zero3_enabled
+from .deepspeed import deepspeed_init, deepspeed_load_checkpoint
 from .dependency_versions_check import dep_version_check
 from .hyperparameter_search import ALL_HYPERPARAMETER_SEARCH_BACKENDS, default_hp_search_backend
 from .modelcard import TrainingSummary
@@ -2742,43 +2742,25 @@ class Trainer:
                 self._save(output_dir, state_dict=state_dict)
             if self.is_fsdp_enabled:
                 if self.args.should_save:
-                    filename = SAFE_WEIGHTS_NAME if self.args.save_safetensors else WEIGHTS_NAME
-                    file = os.path.join(output_dir, filename)
-                    if os.path.isfile(file):
-                        os.remove(file)
+                    for filename in [WEIGHTS_NAME, SAFE_WEIGHTS_NAME]:
+                        file = os.path.join(output_dir, filename)
+                        if os.path.isfile(file):
+                            os.remove(file)
                 save_fsdp_model(self.accelerator.state.fsdp_plugin, self.accelerator, self.model, output_dir)
 
         elif self.is_deepspeed_enabled:
             # this takes care of everything as long as we aren't under zero3
-            if self.args.should_save:
-                state_dict = None
-                if not is_deepspeed_zero3_enabled():
-                    if version.parse(accelerate_version) <= version.parse("0.20.3"):
-                        raise ValueError("Install Accelerate from main branch")
-                    state_dict = self.accelerator.get_state_dict(self.deepspeed)
+            if version.parse(accelerate_version) <= version.parse("0.20.3"):
+                raise ValueError("Install Accelerate from main branch")
+            try:
+                state_dict = self.accelerator.get_state_dict(self.deepspeed)
                 self._save(output_dir, state_dict=state_dict)
-
-            if is_deepspeed_zero3_enabled():
-                # It's too complicated to try to override different places where the weights dump gets
-                # saved, so since under zero3 the file is bogus, simply delete it. The user should
-                # either use deepspeed checkpoint to resume or to recover full weights use
-                # zero_to_fp32.py stored in the checkpoint.
-                if self.args.should_save:
-                    file = os.path.join(output_dir, WEIGHTS_NAME)
-                    if os.path.isfile(file):
-                        # logger.info(f"deepspeed zero3: removing {file}, see zero_to_fp32.py to recover weights")
-                        os.remove(file)
-
-                # now save the real model if stage3_gather_16bit_weights_on_model_save=True
-                # if false it will not be saved.
-                # This must be called on all ranks
-                if not self.model_wrapped.save_16bit_model(output_dir, WEIGHTS_NAME):
-                    logger.warning(
-                        "deepspeed.save_16bit_model didn't save the model, since"
-                        " stage3_gather_16bit_weights_on_model_save=false. Saving the full checkpoint instead, use"
-                        " zero_to_fp32.py to recover weights"
-                    )
-                    self.model_wrapped.save_checkpoint(output_dir)
+            except ValueError:
+                logger.warning(
+                    " stage3_gather_16bit_weights_on_model_save=false. Saving the full checkpoint instead, use"
+                    " zero_to_fp32.py to recover weights"
+                )
+                self.model_wrapped.save_checkpoint(output_dir)
 
         elif self.args.should_save:
             self._save(output_dir)

--- a/src/transformers/trainer.py
+++ b/src/transformers/trainer.py
@@ -2754,7 +2754,8 @@ class Trainer:
                 raise ValueError("Install Accelerate from main branch")
             try:
                 state_dict = self.accelerator.get_state_dict(self.deepspeed)
-                self._save(output_dir, state_dict=state_dict)
+                if self.args.should_save:
+                    self._save(output_dir, state_dict=state_dict)
             except ValueError:
                 logger.warning(
                     " stage3_gather_16bit_weights_on_model_save=false. Saving the full checkpoint instead, use"

--- a/src/transformers/trainer.py
+++ b/src/transformers/trainer.py
@@ -2737,21 +2737,25 @@ class Trainer:
             or self.fsdp is not None
             or self.is_fsdp_enabled
         ):
+            state_dict = self.model.state_dict()
+            if self.args.should_save:
+                self._save(output_dir, state_dict=state_dict)
             if self.is_fsdp_enabled:
-                os.makedirs(output_dir, exist_ok=True)
-                save_fsdp_model(self.accelerator.state.fsdp_plugin, self.accelerator, self.model, output_dir)
-            else:
-                state_dict = self.model.state_dict()
-
                 if self.args.should_save:
-                    self._save(output_dir, state_dict=state_dict)
+                    filename = SAFE_WEIGHTS_NAME if self.args.save_safetensors else WEIGHTS_NAME
+                    file = os.path.join(output_dir, filename)
+                    if os.path.isfile(file):
+                        os.remove(file)
+                save_fsdp_model(self.accelerator.state.fsdp_plugin, self.accelerator, self.model, output_dir)
+
         elif self.is_deepspeed_enabled:
             # this takes care of everything as long as we aren't under zero3
-            if self.args.should_save and not is_deepspeed_zero3_enabled():
-                if version.parse(accelerate_version) <= version.parse("0.20.3"):
-                    raise ValueError("Install Accelerate from main branch")
-                state_dict = self.accelerator.get_state_dict(self.deepspeed)
-
+            if self.args.should_save:
+                state_dict = None
+                if not is_deepspeed_zero3_enabled():
+                    if version.parse(accelerate_version) <= version.parse("0.20.3"):
+                        raise ValueError("Install Accelerate from main branch")
+                    state_dict = self.accelerator.get_state_dict(self.deepspeed)
                 self._save(output_dir, state_dict=state_dict)
 
             if is_deepspeed_zero3_enabled():

--- a/src/transformers/trainer.py
+++ b/src/transformers/trainer.py
@@ -2741,11 +2741,6 @@ class Trainer:
             if self.args.should_save:
                 self._save(output_dir, state_dict=state_dict)
             if self.is_fsdp_enabled:
-                if self.args.should_save:
-                    for filename in [WEIGHTS_NAME, SAFE_WEIGHTS_NAME]:
-                        file = os.path.join(output_dir, filename)
-                        if os.path.isfile(file):
-                            os.remove(file)
                 save_fsdp_model(self.accelerator.state.fsdp_plugin, self.accelerator, self.model, output_dir)
 
         elif self.is_deepspeed_enabled:

